Refactor feature envy in machineShopSimulator

isaacyoa opened the pull request:
We refactored changeState method in machineShopSimulator which included moving chunks of the code to machine in new methods.

isaacyoa committed:
Refactor changeState method

We encapsulated `timeNow` and `largeTime` so
we could move logic from `changeState` to
new methods in Machine called `jobReady`
and `runJob`. This cleans some of the
Feature Envy smell from this code.

isaacyoa committed:
Add comments to explain refactor changes

We added the comments above the methods
'jobReady()' and 'runJob()', and in 'changeState'
above where 'jobReady()' is called.

miyat001 commented:
The issue is well documented. You explain which function is problematic and what you can improve by this change. In pull request, you should explain more about which code you moved instead of called chunks of code. In addition, you should explain which code you created in Machine.java. Your commit message is great. You explain which function and variable are moved. If you ask me, you should write more about getTimeNow, setTimeNow, and getLargeTime.

The refactoring is very clear and complete. As you planned, you moved the change-state function to outside of Machine Shop Simulator.  You create new function "jobReady" in Machine class. It makes code shorter than before. Moreover, you create new functions: getTimeNow, setTimeNow, getLargeTime in Machine Shop Simulator. Your changes will surely improve the condition of code.

bemica commented on .idea/modules/Machine-shop-simulator_test.iml:
```diff
@@ -3,7 +3,9 @@
-    <content url="file://$MODULE_DIR$/../../src/test" />
+    <content url="file://$MODULE_DIR$/../../src/test">
+      <sourceFolder url="file://$MODULE_DIR$/../../src/test/resources" type="java-test-resource" />
+    </content>
```
Remove these .iml files from the pull request

CMenne replied:
+1

bemica commented on src/applications/Machine.java:
```diff
@@ -15,6 +15,23 @@
+        totalWait = totalWait + MachineShopSimulator.getTimeNow() - activeJob.getArrivalTime();
+        numTasks++;
+    }
+
```
I really like adding jobReady() and runJob() as methods, makes the logic of changeState() way more understandable

bemica reviewed (approved):
I'm not convinced that timeNow and largeTime need to be encapsulated, but I'm not necessarily opposed.

Glad you add some much needed documentation

Moving a lot of the logic from changeState() to Machine makes that method *way* more readable

Remove the .iml files from the pull request, but apart from that it's great

CMenne commented on src/applications/MachineShopSimulator.java:
```diff
@@ -26,50 +26,40 @@
-                machine[theMachine].setNumTasks(machine[theMachine].getNumTasks() + 1);
-                int t = machine[theMachine].getActiveJob().removeNextTask();
-                eList.setFinishTime(theMachine, timeNow + t);
-            }
```
I really like how you guys refactored this logic of changeState() out. It really cleans of the readability of the method. :+1:

CMenne reviewed (approved):
I feel that your pull request could be documented with your changes a little more. It provides the basics of what you did which is good, but I feel that some more specific details of what you've done could have been helpful. Your commit messages look detailed and explain the changes you've made nicely, good job.

isaacyoa committed:
Update `.gitignore` to ignore `*.iml` files

We updated `.gitignore` so it will now ingore
files with the `.iml` ending. We also removed
the two `.iml` files we had already committed.

leonidscott commented on .gitignore:
```diff
@@ -66,9 +66,9 @@ fabric.properties
-# .idea/misc.xml
+ *.iml
+ modules.xml
+ .idea/misc.xml
```
:thumbsup:

leonidscott commented on src/applications/Machine.java:
```diff
@@ -14,6 +14,27 @@
+    // jobReady checks to see if there is a job in the queue
+    // and if there is, goes to runJob() and returns time
+    // it took to finish
+    public int jobReady() {
```
I think this function could use a little work: 
* First, I would say that its name `jobReady()` is framed as a question. If I were to see that name, I would assume it would return a boolean. Instead, it actually performs a bunch of side affects. It should really be named more as an action, for example `startNextJob()`.  
* At first glance, I was pretty worried that the two return statements in the function, while both ints, really meant different things, but looking at it again, it doesn't look so bad.  
* That is a weirdly phrased if statement. I would recommend adding the `{}'s` back in. That way you don't need those weird `()'s` around your return statement.  
* The comments really don't add anything. Comments should not be about implementation for several reasons:  
  * It just mimics what the code is saying, and the code can inherently say it better.  
  * People generally don't look at comments above a function declaration.  
  * When this function gets changed next, you also have to worry about updating all the comments. If you, or whomever changes this next *doesn't* do this, those comments become lies.  

   This isn't saying there shouldn't be comments here, just that they should focus more on *what* the code does, rather than *how*. In this case, the logic is very simple, so I wouldn't even bother putting any comments in here.  

There are somethings that I really do like about this function though. First and foremost, I like the `MachineShopSimulator.getLargeTime()`. It is something that should have been in this code base long ago. However, I'm not so sure about the name `.getLargeTime()`. Without looking at MachineShopSimulator, I have *no* idea what it does.

leonidscott commented on src/applications/Machine.java:
```diff
@@ -14,6 +14,27 @@
+    // runJob removes the current job frome the queue
+    // and increments totalWait by the timeNow subtracted
+    // by the arrival time
+    public void runJob() {
```
It seems as if this function is only called once as a helper function to `jobReady()`. Considering this, it really needs to be `private`.  

A similar comment about the use of comments above this method as before. I would say cut them loose.

leonidscott commented on src/applications/MachineShopSimulator.java:
```diff
@@ -26,50 +26,40 @@
     static boolean moveToNextMachine(Job theJob, SimulationResults simulationResults) {
         if (theJob.getTaskQ().isEmpty()) {// no next task
-            simulationResults.setJobCompletionData(theJob.getId(), timeNow, timeNow - theJob.getLength());
+            simulationResults.setJobCompletionData(theJob.getId(), getTimeNow(), getTimeNow() - theJob.getLength());
```
`getTimeNow()` refers to a field that this function has access to. It should just call `timeNow` instead.

leonidscott commented on src/applications/MachineShopSimulator.java:
```diff
@@ -26,50 +26,40 @@
             // put on machine p's wait queue
             machine[p].getJobQ().put(theJob);
-            theJob.setArrivalTime(timeNow);
+            theJob.setArrivalTime(getTimeNow());
```
getTimeNow() refers to a field that this function has access to. It should just call timeNow instead.

leonidscott commented on src/applications/MachineShopSimulator.java:
```diff
@@ -26,50 +26,40 @@
+            theJob.setArrivalTime(getTimeNow());
             // if p idle, schedule immediately
-            if (eList.nextEventTime(p) == largeTime) {// machine is idle
+            if (eList.nextEventTime(p) == getLargeTime()) {// machine is idle
```
`getLargeTime()` referes to a field that this function already has access to. It should just directly refer to `largeTime`.

leonidscott commented on src/applications/MachineShopSimulator.java:
```diff
@@ -26,50 +26,40 @@
 
     /**
-     * change the state of theMachine
+     * change the state of theMachineJobCom
```
I don't know what `theMachineJobCom` is supposed to mean. This is the first time it is referred to in this class. Moreover, you have inherited an awful naming scheme in this code base from the people who originally wrote it. The *the* in `theBlah` adds no meaning, and is no way shape or form a standard java convention. The history of the code base is not your fault ... but it is your problem. It is all of our responsibilities to leave each bit of code we find at least a little bit better than we found them, even if it is not our refactoring mission to fix them, and we should not persist bad naming conventions just because it is not exactly one of our issues.  

I would name this `machineJobCom`, and next to it describe what the *JobCom* part of that means.

leonidscott commented on src/applications/MachineShopSimulator.java:
```diff
@@ -26,50 +26,40 @@
     static Job changeState(int theMachine) {// Task on theMachine has finished,
                                             // schedule next one.
-        Job lastJob;
+        Job lastJob = null;
```
`lastJob` should default to `null` if it is initialized but not defined. I would just leave it as `Job lastJob`.

leonidscott commented on src/applications/MachineShopSimulator.java:
```diff
@@ -26,50 +26,40 @@
-                eList.setFinishTime(theMachine, timeNow + t);
-            }
+            //jobReady returns time it took to finish job
+            eList.setFinishTime(theMachine, machine[theMachine].jobReady());
```
The refactoring you did here certainly helps readability and removes a message chain. That is a big plus! 

However, you can see how the name `.jobReady()` does not work here. The name `jobReady()` indicates a *state*, not an *action*. It should be renamed that there is no doubt that when you call it, that it is making side effects and not just returning a boolean. Moreover, the name mentions nothing about the fact that `.jobReady()` returns a *time*. This is especially an eye sore in a function that very clearly deals with time from its name.

leonidscott commented on src/applications/MachineShopSimulator.java:
```diff
@@ -26,50 +26,40 @@
             lastJob = machine[theMachine].getActiveJob();
             machine[theMachine].setActiveJob(null);
-            eList.setFinishTime(theMachine, timeNow
+            eList.setFinishTime(theMachine, getTimeNow()
```
`getTimeNow()` refers to a field that this function has access to. It should just call `timeNow` instead.

leonidscott commented on src/applications/MachineShopSimulator.java:
```diff
@@ -104,7 +94,7 @@ private static void setUpJobs(SimulationSpecification specification) {
     private static void createEventAndMachineQueues(SimulationSpecification specification) {
         // create event and machine queues
-        eList = new EventList(specification.getNumMachines(), largeTime);
+        eList = new EventList(specification.getNumMachines(), getLargeTime());
```
`getLargeTime()` refers to a field that this function already has access to. It should just directly refer to `largeTime`.

leonidscott commented on src/applications/MachineShopSimulator.java:
```diff
@@ -133,7 +123,7 @@ static void startShop(SimulationSpecification specification) {
         while (numJobs > 0) {// at least one job left
             int nextToFinish = eList.nextEventMachine();
-            timeNow = eList.nextEventTime(nextToFinish);
+            setTimeNow(eList.nextEventTime(nextToFinish));
```
You are using a method to set access to a variable you already have access to. You have access to that variable so you should just set it.

leonidscott commented on src/applications/MachineShopSimulator.java:
```diff
@@ -146,7 +136,7 @@ static void simulate(SimulationResults simulationResults) {
      * @param simulationResults*/
     static void outputStatistics(SimulationResults simulationResults) {
-        simulationResults.setFinishTime(timeNow);
+        simulationResults.setFinishTime(getTimeNow());
```
`getTimeNow()` refers to a field that this function has access to. It should just call `timeNow` instead.

leonidscott commented on src/applications/MachineShopSimulator.java:
```diff
@@ -170,7 +160,7 @@ private static void setNumTasksPerMachine(SimulationResults simulationResults) {
     public static SimulationResults runSimulation(SimulationSpecification specification) {
         largeTime = Integer.MAX_VALUE;
-        timeNow = 0;
+        setTimeNow(0);
```
You are using a function to set a variable you already have access to. You should just set that variable directly.

leonidscott commented on src/applications/MachineShopSimulator.java:
```diff
@@ -191,4 +181,16 @@ public static void main(String[] args) {
         simulationResults.print();
     }
+
+    public static int getTimeNow() {
```
I would argue that this function and `setTimeNow` actually harm the readability of your code. The aim of making a variable private is to limit its possible interactions with the outside world. However, having a setter and getter that give full access to your variable anytime, anywhere allows anyone to access that variable as if it was public ... so why not just make it public. It has a need to be read and written to, so let it be.

leonidscott commented on src/applications/MachineShopSimulator.java:
```diff
@@ -191,4 +181,16 @@ public static void main(String[] args) {
+        MachineShopSimulator.timeNow = timeNow;
+    }
+
+    public static int getLargeTime() {
```
Because this function makes `largeTime` readable, but not writable. This getter function makes it harder to make a mistake with `largeTime`. :thumbsup:

leonidscott reviewed (commented):
Your function, `jobReady()`, in Machine executes your goal of reducing feature envy well. It makes the structure of the code in MachineShopSimulator much better. In addition, your issues are well documented.

However there are several things in this codebase that I would consider *blocking problems* that should be addressed before merging into master.  
* **Getters and Setters for variables `timeNow`, and `largeTime` in MachineShopSimulator:** There are three problems I see with how this is implemented that makes the code *harder* to understand:  
  * `timeNow` is set as a private variable. This should be done to limit the interactions between `timeNow`, and the outside world. However, the fact that there are getters and setters that let any part of the project do whatever they want with this variable completely defeats the point of `timeNow` being private. `timeNow` is a very extensively used variable and it needs to be set in many different contexts; so let it. Going through these extra functions only adds a false sense of security for the user, and adds complexity to the code.  
  * The setters and getters are used in a class where the variables `timeNow` and `longTime` are already accessible by any function. Why complect these parts of the code by having them call a function that just re-routes them to a variable they already have access to?  
  * The variable `longTime` is not named very well. It isn't your fault... but it is your problem. The comment  right next to its declaration says *"*`all machines finish before this`*"*, but the name makes it seem like it is the same as timeNow, but a long? It would instead make sense to name it something like `endTime`. We should always carry the responsibility of making the code we work on just that little bit better than when we found it, even if it is not necessarily what we set out to refactor.   
* **`Machine.jobReady()`:** This function name seems like it has something to do with *state* of some object, not that it does an *action*. It should be named in a way that conveys that, such as `runJobAndSetTime()` (although that probably too long of a variable name). Moreover, it returns an int representing *time*, but the name gives no hints about this, and it really should.  
* **Commit Messages:** All commits should refer to an issue that they are fixing. You can do this by adding a `[#<IssueNumber>]` inside your commit message. This way, when you pull up an issue on github, you can see all the commits that pertain to that issue.  

I would highly recommend addressing these issues before moving too far along. The more changes are made ontop of this PR, the harder it will be to fix these things later.
